### DISCUSSION TITLE
Performance of eval() in freezed apps

### DISCUSSION THREAD

**Discussion by mbllo-gh** [marcelotduarte/cx_Freeze#1206] [Q&A]
I created an app that is using eval() for deserialization. Unfortunately, the freezed program takes 6-10 x more time to execute
eval() statements than the unfreezed which makes the start of my app quite painful. Any idea why that happens and how to fix?
Thanks.

**Comment by marcelotduarte** [comment]
What usually makes a frozen program slow is putting the packages in the zip. I think the eval itself must be slow.
What is the structure of the program? If you can show me an example, so I can suggest something.

**Comment by mbllo-gh** [comment]
While trying to reduce amount of code for a testing example I could enclose the problem even more.
The eval() statement is executing object constructors and some of them are using inspect.stack() which is the main performance killer at all.
In my own code it needs 2-3ms per call, freezed then 20-30ms.

To demonstrate I created a small test:

import inspect
import timeit

def subsub(n: int):
    start = timeit.default_timer()
    frames = inspect.stack()
    end = timeit.default_timer()
    print("=====", n, (end - start) * 1000)

def sub(n: int):
    if n < 20:
        n += 1
        sub(n)
    else:
        subsub(n)

for i in range(10):
    sub(0)


And setup.py:

from cx_Freeze import setup, Executable
executables = []
executables.append(Executable("test_inspect.py"))
setup(name="test_inspect",
      executables=executables)

**Comment by mbllo-gh** [comment]
While trying to reduce amount of code for a testing example I could enclose the problem even more.
The eval() statement is executing object constructors and some of them are using inspect.stack() which is the main performance killer at all.
In my own code it needs 2-3ms per call, freezed then 20-30ms.

To demonstrate I created a small test:

import inspect
import timeit

def subsub(n: int):
    start = timeit.default_timer()
    frames = inspect.stack()
    end = timeit.default_timer()
    print("=====", n, (end - start) * 1000)

def sub(n: int):
    if n < 20:
        n += 1
        sub(n)
    else:
        subsub(n)

for i in range(10):
    sub(0)


And setup.py:

from cx_Freeze import setup, Executable
executables = []
executables.append(Executable("test_inspect.py"))
setup(name="test_inspect",
      executables=executables)

**Comment by mbllo-gh** [comment]
While trying to reduce amount of code for a testing example I could enclose the problem even more.
The eval() statement is executing object constructors and some of them are using inspect.stack() which is the main performance killer at all.
In my own code it needs 2-3ms per call, freezed then 20-30ms.

To demonstrate I created a small test:

import inspect
import timeit

def subsub(n: int):
    start = timeit.default_timer()
    frames = inspect.stack()
    end = timeit.default_timer()
    print("=====", n, (end - start) * 1000)

def sub(n: int):
    if n < 20:
        n += 1
        sub(n)
    else:
        subsub(n)

for i in range(10):
    sub(0)


And setup.py:

from cx_Freeze import setup, Executable
executables = []
executables.append(Executable("test_inspect.py"))
setup(name="test_inspect",
      executables=executables)
